### ISSUE TITLE
package and module private visibility

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#1128]
I think we all know now that plain `shared`, as nice as it is, doesn't account for all the use cases that library designers run into, and that our current visibility model adds unnecessary design constraints. I think the following syntax for restricting the visibility of a `shared` program element is reasonable:

``` ceylon
shared (`package`) //package visibility
shared (`module`) //module visibility
```

(This option was opened up by #1115.)

Of course, this approach could also accommodate a list of packages/modules:

``` ceylon
shared (`package`, `package other.friendly.package`)
class ALittleBitShared() {}

shared (`module`, `module tests`)
package testable;
```

Of course this stuff is going to have to come with some limitations. I'm thinking:
- This always _reduces_ the impact of `shared`, that is, you can't use it to make a program element visible where its container is not visible.
- Refinement can widen the visibility of a member, but it can't narrow the visibility of a member.

Thoughts?

**Comment by rbi** [comment]
I know ceylon only from some tutorials and therefore may not be familiar with all its concepts. The main reason for not looking further into ceylon is the missing support for class level test for non-shared classes/packages.

Wouldn't it be possible to support some kind of `merge import` or `module extension`? In Maven you have usually two Java source folders. One for the productive code and one for the tests. Its a common pattern to have the same package hierarchy in both of them so that the test classes are in the same package as the productive classes. This way package private classes can be tested without problems. I really like this concept because a test class for one specific class is per definition not reusable and therefore it makes no sense to have it anywhere else as in the same package of the class it tests. Not deploying the test classes at runtime is mostly an optimization measure.

Maybe in ceylon it would be nice to have such "special" source folders too which allow to provide extensions to other modules by duplicating the package hierarchy e.g.:

~~~
|-source
|  |-somemodule
|     |-module.cylon
|     |-SomeClass.cylon
|-test-source
|  |-somemodule
|     |-module-extension.cylon
|     |-SomeClassTest.cylon
~~~

No softening of the visibility modifiers would be necessary. The extension would augment the original module by some classes and dependencies and produce a separate version of the module with something like a Maven classifier.

**Comment by quintesse** [comment]
@rbi It's already possible to do this with the current compiler although it would be somewhat of a hack. First of all you can just add several source folders, so you could compile the above with `ceylon compile --src source --src test-source somemodule` and already get *almost* what you want.

The problem is the `module-extension.ceylon` file. You could do something like this:

~~~
|-source
|  |-somemodule
|     |-SomeClass.ceylon
|-module-source
|  |-somemodule
|     |-module.ceylon
|-test-source
|  |-somemodule
|     |-module.ceylon
|     |-SomeClassTest.ceylon
~~~

and compile with `ceylon compile --src source --src test-source somemodule` for tests and with `ceylon compile --src source --src module-source somemodule` for release.

But like I said, that's a bit of a hack. In an earlier version of the compiler I had it actually set up to allow duplicate source files and the compiler would only look at the first one it encountered. So you could have the files setup like this:

~~~
|-source
|  |-somemodule
|     |-module.ceylon
|     |-SomeClass.ceylon
|-test-source
|  |-somemodule
|     |-module.ceylon
|     |-SomeClassTest.ceylon
~~~

and compile with `ceylon compile --src test-source --src source somemodule` (see that the 'test-source' is mentioned first) and it would just have worked. But unfortunately that was removed in favor of disallowing duplicates.

**Comment by rbi** [comment]
@quintesse thanks for the information. I tried it out in the Ceylon IDE and it seams to work reasonable well when the module-source folder is not included in eclipse. I can than build the release archive with the command line you've given in your example but only if i delete the generated "module" directory. If this is not done, than the .car build on the command line still contains the test classes.

So with some additional build-script magic this could work.

**Comment by quintesse** [comment]
@rbi yes, for the test module you should probably do something like: `ceylon compile --src source --src test-source --out test-modules somemodule` so your test modules get stored separately from the release modules.

**Comment by FroMage** [comment]
Jigsaw has gone with friend modules. I think it makes no sense to make the friend modules visible unless imported too. I see these concepts as orthogonal.

**Comment by akberc** [comment]
Just to re-iterate a comment I made somewhere a couple of years ago : 'Test' modules or packages or compiling to the same package to get access would be implementing Java's solution based on its visibility scheme, and not native to Ceylon and not in tune with the central role of testing in today's software development.

Why not compile members annotated 'test' when actually running tests, and Test classes or methods can exist anywhere in the same package - in a separate compilation unit, or even in the same unit.  I find it most natural when writing Ceylon to write the test in the same compilation unit.  In fact, separating classes and compilation units as Ceylon has done, naturally leads to this approach, where you put a class, its associated aliases, tests etc. in one compilation unit.

All the compiler needs to do is to compile tests only when requested : `--include-tests` and when invoked by the tester due to uncompiled tests.

In short, any flexibility in the current visibility scheme may have unintended security and 'innovation' consequences down the road, and there are much more natural and modern ways to handle this issue, which is indeed an impediment to TDD.

**Comment by tombentley** [comment]
>where you put a class, its associated aliases, tests etc. in one compilation unit.
>
>All the compiler needs to do is to compile tests only when requested : --include-tests and when invoked by the tester due to uncompiled tests.

I do find the thought of having the tests right next to the thing they're testing pretty attractive. Not having to switch between editors would be a definite plus, and I think people would be more likely to actually write tests. But it's not as simple as you're suggesting. The `test`-annotated methods might call utility functions which should not be part of a non-test `car`, likewise the tests might depend on modules which that non-test module should not have as dependencies.

**Comment by quintesse** [comment]
It's a bit of a crazy idea, but all the basic work for this is now provided for by the `native` annotations. They allow you to mark code and imports as being only required for a single backend. Theoretically we could create something similar, almost like having a "test" backend. So it does seem possible at least.

**Comment by gavinking** [comment]
@akberc to be honest I'm pretty sympathetic to this suggestion, and I agree that it better reflects the real-life workflow. I agree that tests are not "really" a separate module. The main problem would be deciding how `test` code relates to artifacts (i.e. packages). I wonder if we could compile stuff annotated `test` into separate artifacts (e.g. `module-version-tests.car`, `module-version-tests.js`) and have the CMR know when not to load them.

**Comment by lucono** [comment]
@gavinking I think the suggested approach of being able to share with specific packages and modules is very powerful and flexible.

My only concern with this approach is that it would allow program elements to make absolute (ie, non-relative) reference to packages/modules as share targets. Such a referenced package, for instance, could be refactored to change in meaning within the module, like broadening or narrowing its functional scope, which would easily cause the `shared` references in the previous context to inadvertently remain undiscovered.

So for instance, I have a class in one package that is shared only with another specifically named package, but later rename and refactor the target package to something functionally different within the module, so that it no longer makes sense that the shared class in the other package should remain shared with the refactored package.

The plain `shared` as it currently exists, is nice and simple, but is limited because of its inability to restrict sharing of program elements to more local outer scopes when the outer element is also shared. The next simplest extension to ease that limitation might be a kind of `shared local` which declares a shared program element to be shared only within the local immediate outer/external scope or parent, without being shared any wider even if the external scope/parent itself is also shared.

So for instance in the following:

```js
// Top-level class in shared package
shared class Agent() {

    shared void start() {
        // code
    }
    
    shared(local) void activate() {
        // code
    }
}
```

The `start` method of the `Agent` class in the above example would be shared and accessible from outside of the module, but the `activate` method would be shared only within the package even though the `Agent` class itself as well as the containing package are also shared.

This `shared local` approach would be a little better than the current simple `shared` approach, but would still not be able to allow, for instance, sharing of the `activate` method of the `Agent` class at the module level, since the immediate "local" sharing scope here is the package, and not the module.

Finally, there's yet another option in my opinion, which is one in which rather than just having a `local` option to shared program elements, it can be explicitly stated to be for the containing package or module. This would be a subset of the current recommendation for being able to reference any absolute package or module (or list of) as a sharing target, but with the important difference that it would avoid absolute references to specific packages/modules as share targets and only support relative outer (ie, containing) package or module share targets.

It could be implemented to leverage #1115, but enforce that only outer package and module can be provided in `shared(...)`. That makes it possible for it to be easily relaxed in the future to also allow specific named packages or lists of packages/modules if practical use over time comes to support that to also be necessary. By then, there may also be new problems as well as ideas from practical use, on how to best evolve the feature.

So my personal opinion is that for now, the best option might be:

```js
shared ... // Same as current
shared(`package`) ... // Shared with outer package
shared(`module`) ... // Shared with outer module

// But not:
shared(`package other.friendly.package`) ... // Does not compile. May compile in future
shared(`module other.friendly.module`) ... // Does not compile. May compile in future
```

**Comment by gavinking** [comment]
@lucono well that is already part of the current proposal, and it falls out naturally, because we already have metamodel refs like `` `package` `` which always refers to the containing package, `` `module` `` which always refers to the containing module, and even ` ``class`` ` which refers to the containing class.

**Comment by luolong** [comment]
@gavinking so I understand it correctly that ``shared( `class` )`` will mean something along the lines of `protected` visibility of Java language?

**Comment by lucono** [comment]
@gavinking Indeed. What I do not like about the current recommendation is support for sharing with absolutely referenced packages/modules as share targets. I think that could cause issues when refactoring large projects (which I explain in my original post), the kind of issue which Ceylon has been great at trying to avoid so far.

I think it would be great to limit it to just the outer package and module as a first step (which already provides significantly greater flexibility than the current plain `shared` implementation), and see how that works out over time. And then only go further to allow sharing with any absolutely referenced packages/modules if practical use of just the outer package/module option seems to support that direction.

**Comment by gavinking** [comment]
@luolong Oops, no, actually now I realize that `` shared(`class`) `` is a synonym for `shared`, so not useful at all actually.

**Comment by luolong** [comment]
@lucono while I agree that generally this is a feature that can cause problems if misused, but there are some quite legitimate use cases for allowing shared access to specific named modules/packages.

Most prominent example of such _"friend"_ module is unit testing.

**Comment by gavinking** [comment]
@lucono well it's really just for test code (some people like to write their unit tests in a different package) and for some really special rare cases. I don't think anyone would advocate the use of absolute references as a thing you do every day. And the syntax for them will be unwieldy enough that people won't want to use them.

**Comment by lucono** [comment]
@gavinking @luolong Maybe Ceylon could have a different way specifically designed for supporting test modules and unit-testing in general. Having production code directly aware of and referencing test packages and modules does feel a little uncomfortable.

It would be great to have some mechanism designed specifically for testing in Ceylon, where it's not necessarily trying to piggy-back on other language features which may not necessarily fit elegantly with both good application design and the ease and ability for full application testing.

Without having given it too much thought, it could be something at a module level, where a module could declare its "test module/s", for which Ceylon could then somehow support automatically relaxed visibility to program elements when being accessed from within the test module.

It could simply be a "friend" designation that is provided by the compiler at compile time based on the application module's declared "test module/s", but something that would not be available to application code to name individual absolutely referenced packages/modules as "share-friends".
